### DISCUSSION TITLE
Memory Usage

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#19660] [Q&A General] [answered]
Hello,

So I am running into memory problems when running a larger mesh code with a lot of variables (almost a 100 variable).

1) Does using `LAGRANGE_VEC` for field data instead of 3 `LAGRANGE` variables would be more efficient memory usage?
2) I am using a userObject in one of the primary kernels that I am using and it has large matrices , this one is called in `computeResidual`, `computeJacobian`, and `computeOffDiagJacobian` functions. for a nodal variable, does this mean that a userObject is created for each node or each element?
3) When using meshes of 1+million elements, it hogs all the memory, is there a way to reduce memory usage? I am already using `--use-distributed-mesh`
4) I can see moose has a memory logger, but it doesnt have a detailed usage for the kernels or objects in the MooseApp I have built, is there a simple way to analyze this? I have tried memscape and most of the tracebacks were leading back to libmesh.

Thanks in advance!

**Comment by GiudGiud** [comment, marked as answer]
Hello

1. Not really any advantage in memory. There's some advantages in simulation time with cache efficiencies when using vector variables. No real difference in how much data has to be stored. @lindsayad might know better
2. No the user object is created once per process. It's not created for every node or element. It may be ran on every node or element, depending on the type of the user object. 
3. To reduce your memory footprint, please have a look at this post:
https://github.com/idaholab/moose/discussions/19460
I think you should work on the numerics / petsc options now to reduce the memory costs. 
4. You can use 
```
[Debug]
  show_actions = true
[]
``` 
to see more information about the memory cost of each object. Note that a kernel or a UO usually do not have a large cost. It's more the cost of storing variables or solving the numerical systems. 

Guillaume

**Reply by lindsayad** [reply]
We have a section on memory profiling [here](https://mooseframework.inl.gov/application_development/profiling.html). Also MOOSE will not recognize the option `--use-distributed-mesh`. It's just `--distributed-mesh`. When MOOSE prints out its header at the beginning of a simulation you should be able to see whether you're replicated or distributed.

I don't think there's much resource advantage between multiple `LAGRANGE` variables and a `LAGRANGE_VEC` variable. The advantage to me is cleaner code and a cleaner input file.

**Reply by KhaledNabilSharafeldin** [reply]
Thank you for the replies and clarifications! 
I am already using -ilu flag for PETSc, it doesn't seem to be that different. 
I'm also using 2 multiapps, so it is probably a big factor in this since i'm transfer-copying a lot of variables.

is there a way to operate on the same variables without creating copies of them?
Also, is there a way to disable restart and restore? if it saves some memory, this might be helpful to test larger problems. it seems to be the largest factor in show actions.

**Reply by lindsayad** [reply]
did you see a change in memory usage between replicated and distributed? I don't know a whole lot about the memory tracking through the actions. I usually rely on the heap profiler to show me exactly what is consuming memory otherwise my guesses are often inaccurate. If you have a MWE, then I could run the profiler for you if you're not well setup to run it yourself
